lib/uknetdev: Provide `uk_netbuf_dup_single`

skuenzer opened the pull request:
This PR introduces `uk_netbuf_dup_single()` that creates a copy of a given single netbuf (out of a chain).

The PR depends on PR #308.

skuenzer committed:
lib/uknetdev: Provide `uk_netbuf_dup_single()`

Introduces `uk_netbuf_dup_single()` that creates a copy
of a given single netbuf (out of a chain).

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

craciunoiuc commented:
Hey @skuenzer,

I did an initial pass through the PR and I couldn't find anything out of the ordinary.
I suppose after #308 gets merged this will need to be rebased and will only contain the last commit.
I'll come back to this then and test it.

craciunoiuc reviewed (approved):
I'll add my review tag now, but this PR depends on #308 to be merged and will also need to be rebased afterwards.

@nderjung it's on you to be careful when adding your approval 😉 

The PR overall looks good, it should be okay.

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>

nderjung reviewed (changes requested):
Hi @skuenzer, #308 has now been merged and this needs to be rebased. :)

nderjung commented:
Migrating to v0.9, this change is too big to review without proper testing.  Suggestion is that the PR should coincide with uknetdev test suite which aims to test this (and the rest) of the library's methods.

skuenzer committed:
lib/uknetdev: Provide `uk_netbuf_poolalloc_buf()`, `uk_netbuf_poolalloc_buf_batch()`

Introduces `uk_netbuf_poolalloc_buf()` and
`uk_netbuf_poolalloc_buf_batch()`, optimized functions that allocate
netbufs directly from a `ukallocpool` memory pool. netbuf size and
buffer alignment are derived from the pool.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uknetdev: Fix size check in `uk_netbuf_prepare_buf()`

Fixes checking that a given memory area has enough space to
also store necessary meta data.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uknetdev: Provide `uk_netbuf_duppool_single()`

Introduces the pool variant of `uk_netbuf_dup_single()`. It creates a
copy of a given single netbuf (out of a chain) and allocates the buffer
from a given `ukallocpool`. Alignment and buffer size are derived from
this pool.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
drivers/virtio: Do not print a warning when a ring is full

Whenever a virtio ring is full we currently print a warning to the
console. Such a situation can happen when I/O is requested faster than
the underlying device can handle (e.g., network packet generator).
Because printing slows further down the processing it may make the
situation worse. This commit turns this warning into a debug message.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>
